metrics/src/main/java/io/opencensus/metrics/TimeSeries.java

```diff
@@ -0,0 +1,157 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.metrics;
+
+import com.google.auto.value.AutoValue;
+import io.opencensus.common.ExperimentalApi;
+import io.opencensus.common.Function;
+import io.opencensus.common.Timestamp;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * A collection of data points that describes the time-varying values of a {@code Metric}.
+ *
+ * @since 0.16
+ */
+@ExperimentalApi
+@Immutable
+public abstract class TimeSeries {
+
+  TimeSeries() {}
+
+  /**
+   * Returns the set of {@link LabelValue}s that uniquely identify this {@link TimeSeries}.
+   *
+   * <p>Apply to all {@link Point}s.
+   *
+   * <p>The order of {@link LabelValue}s must match that of {@link LabelKey}s in the {@code
+   * MetricDescriptor}.
+   *
+   * @return the {@code LabelValue}s.
+   * @since 0.16
+   */
+  public abstract List<LabelValue> getLabelValues();
+
+  /**
+   * Returns the data {@link Point}s of this {@link TimeSeries}.
+   *
+   * @return the data {@code Point}s.
+   * @since 0.16
+   */
+  public abstract List<Point> getPoints();
+
+  /**
+   * Applies the given match function to the underlying data type.
+   *
+   * @since 0.16
+   */
+  public abstract <T> T match(
+      Function<? super TimeSeriesGauge, T> p0,
+      Function<? super TimeSeriesCumulative, T> p1,
+      Function<? super TimeSeries, T> defaultFunction);
+
+  /**
+   * A collection of data points that describes the time-varying values of a gauge {@code Metric}.
+   *
+   * @since 0.16
+   */
+  @ExperimentalApi
+  @Immutable
+  @AutoValue
+  public abstract static class TimeSeriesGauge extends TimeSeries {
+
+    TimeSeriesGauge() {}
+
+    @Override
+    public final <T> T match(
+        Function<? super TimeSeriesGauge, T> p0,
+        Function<? super TimeSeriesCumulative, T> p1,
+        Function<? super TimeSeries, T> defaultFunction) {
+      return p0.apply(this);
+    }
+
+    /**
+     * Creates a {@link TimeSeriesGauge}.
+     *
+     * @param labelValues the {@code LabelValue}s that uniquely identify this {@code TimeSeries}.
+     * @param points the data {@code Point}s of this {@code TimeSeries}.
+     * @return a {@code TimeSeriesGauge}.
+     * @since 0.16
+     */
+    public static TimeSeriesGauge create(List<LabelValue> labelValues, List<Point> points) {
+      // Fail fast on null lists to prevent NullPointerException when copying the lists.
+      Utils.checkNotNull(labelValues, "labelValues");
+      Utils.checkNotNull(points, "points");
```

Comment:
This method should also check the list elements for null.

Reply:
Done. Added a Util method to check list element.